hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -0,0 +1,243 @@
+              + "replicas needed: {}. Number of available source replicas are:"
+              + " {}. Available sources are: {}", container.containerID(),
+              repConfig.getData(), sources.size(), sources);
+        }
+      } else if (containerHealthResult.underReplicatedDueToDecommission()) {
+        Map<DatanodeDetails, SCMCommand> commands = new HashMap<>();
+        Set<Integer> decomIndexes = replicaCount.decommissioningIndexes();
+        final List<DatanodeDetails> selectedDatanodes =
+            getTargetDatanodes(replicas, container, decomIndexes.size());
+        Iterator<DatanodeDetails> iterator = selectedDatanodes.iterator();
+        // In this case we need to do one to one copy.
+        for (ContainerReplica replica : replicas) {
+          if (decomIndexes.contains(replica.getReplicaIndex())) {
+            if (!iterator.hasNext()) {
+              LOG.warn("Couldn't find the enough targets. Available source"
+                  + " nodes: {}, the target nodes: {} and the decommission"
+                  + " indexes: {}", replicas, selectedDatanodes, decomIndexes);
+              break;
+            }
+            DatanodeDetails decommissioningSrcNode
+                = replica.getDatanodeDetails();
+            final ReplicateContainerCommand replicateCommand =
+                new ReplicateContainerCommand(id.getProtobuf().getId(),
+                    ImmutableList.of(decommissioningSrcNode));
+            DatanodeDetails target = iterator.next();
+            commands.put(target, replicateCommand);
+          }
+        }
+        return commands;
+      } else {
+        LOG.info("There are no missing indexes or decommissioning indexes"
+                + " found for {}", id);
+      }
+
+    } catch (IOException | IllegalStateException ex) {
+      LOG.warn("Exception while processing for creating the EC reconstruction" +
+              " container commands for {}.",
+          id, ex);
+    }
+    return null;
+  }
+
+  private List<DatanodeDetails> getTargetDatanodes(
+      Set<ContainerReplica> replicas, ContainerInfo container,
+      int requiredNodes) throws IOException {
+    // We should ensure that the target datanode has enough space
+    // for a complete container to be created, but since the container
+    // size may be changed smaller than origin, we should be defensive.
+    final long dataSizeRequired =
+        Math.max(container.getUsedBytes(), currentContainerSize);
+    final List<DatanodeDetails> excludeList =
+        replicas.stream().map(ContainerReplica::getDatanodeDetails)
+            .collect(Collectors.toList());
+
+    return containerPlacement
+        .chooseDatanodes(excludeList, null, requiredNodes, 0, dataSizeRequired);
+  }
+
+  private static byte[] int2byte(List<Integer> src) {
+    byte[] dst = new byte[src.size()];
+
+    for (int i = 0; i < src.size(); i++) {
+      dst[i] = (byte) ((int) src.get(i));
```

Comment:
Nit: could use `byteValue` directly?

```suggestion
      dst[i] = src.get(i).byteValue();
```

Reply:
ah, done